text/0000-simd-ffi.md

```diff
@@ -0,0 +1,220 @@
+
+And this is where things went wrong: in Rust, `__m256` is always available
+independently of whether `AVX` is available or not<sup>[1](#layout_unspecified)</sup>, 
+but we haven't specified how we are actually compiling our Rust program above:
+
+* if we compile it with `AVX` globally enabled, e.g., via `-C
+  target-feature=+avx`, then the behavior of calling `foo` is defined because
+  `__m256` will be passed to C in a single 256-bit wide register, which is what
+  the C ABI requires.
+  
+* if we compile our program without `AVX` enabled, then the Rust program cannot
+  use 256-bit wide registers because they are not available, so independently of
+  how `__m256` will be passed to C, it won't be passed in a 256-bit wide
+  register, and the behavior is undefined because of an ABI mismatch.
+
+<a name="layout_unspecified">1</a>: its layout is currently unspecified but that
+is not relevant for this issue - what matters is that 256-bit registers are not
+available and therefore they cannot be used.
+
+You might be wondering: why is `__m256` available even if `AVX` is not
+available? The reason is that we want to use `__m256` in some parts of
+Rust's programs even if `AVX` is not globally enabled, and currently we don't
+have great infrastructure for conditionally allowing it in some parts of the
+program and not others.
+
+Ideally, one should only be able to use `__m256` and operations on it if `AVX`
+is available, and this is exactly what this RFC proposes for using vector types
+in C FFI: to always require `#[target_feature(enable = X)]` in C FFI functions
+using SIMD types, where "unblocking" the use of each type requires some
+particular feature to be enabled, e.g., `avx` or `avx2` in the case of `__m256`.
+
+That is, the compiler would reject the example above with an error: 
+
+```
+error[E1337]: `__m256` on C FFI requires `#[target_feature(enable = "avx")]`
+ --> src/main.rs:7:15
+  |
+7 |     fn foo(x: __m256) -> __m256;
+  |               ^^^^^^
+```
+
+And the following program would always have defined behavior
+([playground](https://play.rust-lang.org/?gist=db651d09441fd16172a5c94711b2ab97&version=nightly&mode=debug&edition=2015)):
+
+```rust
+#[target_feature(enable = "avx")]
+extern "C" fn foo(x: __m256) -> __m256;
+
+fn main() {
+    unsafe { 
+        union U { v: __m256, a: [u64; 4] }
+        if is_x86_feature_detected!("avx") {
+            foo(U { a: [0; 4] }.v);
+        }
+    }
+}
+```
+
+independently of the `-C target-feature`s used globally to compile the whole
+binary.
+
+Note here that:
+
+* `extern "C" foo` is compiled with `AVX` enabled, so `foo` takes an `__m256`
+  like the C ABI expects
+* the call to `foo` is guarded with an `is_x86_feature_detected`, that is, `foo`
+  will only be called if `AVX` is available at run-time
+* if the Rust binary is compiled without `AVX`, Rust will insert shims in the
+  call to `foo` to pass it as a 256-bit register. Rust already does this, and
```

Comment:
Does it? I might have missed something, but AFAIK we just pass the vectors in memory, not generate actual ABI-adapting shims (and a quick experiment on the playground seems to confirm that). That strategy doesn't work for extern functions since we can't modify the receiving side's ABI. We'd need to start generating such shims.

Reply:
Honestly, I am not sure what we do here anymore (e.g. see https://github.com/rust-lang/rust/pull/55073 which might be reverted? cc @alexcrichton , https://github.com/rust-lang/rust/pull/55024 ).

But yes, extern functions need to use the target calling convention, which is often (always?) to take vectors by value (see the SysV ABI discussion above). This means that iff the Rust calling convention is to pass vectors by memory (which is not the case for some targets like wasm?), then we need to adapt those calling conventions when doing code generation. 

I think I should leave this part as vague as that and remove the mentions of the "shims". What exactly we have to do here in rustc is an implementation detail.

Reply:
We should pass SIMD types as efficiently as possible in native Rust, which would normally be to do so following the ABI.

Reply:
The problem is which ABI to use on targets with multiple ABIs, like `x86_64`.